### PR TITLE
Switch to persistent adaptive radix trees

### PR DESCRIPTION
**Pull request by joamaki**
Switch StateDB to use custom implementation of Persistent Adaptive Radix Trees (https://db.in.tum.de/~leis/papers/ART.pdf).

The implementation isn't a completely fateful implementation of the paper. The Node16 SIMD tricks and Node48 key array are skipped. This made the implementation simpler, but did leave some performance improvements on the table.

The only functional change to StateDB is to drop the `Last` and `LastWatch` methods since that requires an reverse iterator implementation and since we didn't use them we can just drop them for now.

Benchmark before:

```
goos: linux                                                                                                                                                                                     
goarch: amd64                                                                                                                                                                                   
pkg: github.com/cilium/statedb                                                                                                                                                                  
cpu: Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz                                                                                                                                                  
BenchmarkDB_WriteTxn_1-8                                          307952              4012 ns/op            249227 objects/sec      3080 B/op         53 allocs/op                              
BenchmarkDB_WriteTxn_10-8                                         504811              2373 ns/op            421360 objects/sec      1501 B/op         23 allocs/op                              
BenchmarkDB_WriteTxn_100-8                                        522646              2229 ns/op            448544 objects/sec      1399 B/op         19 allocs/op                              
BenchmarkDB_WriteTxn_1000-8                                       369973              2954 ns/op            338547 objects/sec      1492 B/op         18 allocs/op                              
BenchmarkDB_WriteTxn_10000-8                                      995700              2792 ns/op            358210 objects/sec      1355 B/op         21 allocs/op                              
BenchmarkDB_WriteTxn_100_SecondaryIndex-8                         293709              3843 ns/op            260247 objects/sec      2208 B/op         33 allocs/op                              
BenchmarkDB_RandomInsert-8                                          1032           1192822 ns/op            838348 objects/sec    796015 B/op      12214 allocs/op                              
BenchmarkDB_RandomReplace-8                                          243           4906222 ns/op            203823 objects/sec   2382866 B/op      33380 allocs/op                              
BenchmarkDB_SequentialInsert-8                                       430           2770646 ns/op            360927 objects/sec   1492345 B/op      18256 allocs/op                              
BenchmarkDB_Baseline_SingleRadix_Insert-8                           3109            371203 ns/op           2693945 objects/sec    353662 B/op       5103 allocs/op                              
BenchmarkDB_Baseline_SingleRadix_TrackMutate_Insert-8               3105            373685 ns/op           2676056 objects/sec    353951 B/op       5106 allocs/op                              
BenchmarkDB_Baseline_SingleRadix_Lookup-8                          13185             90097 ns/op          11099219 objects/sec         0 B/op          0 allocs/op                              
BenchmarkDB_Baseline_Part_RootOnlyWatch_Insert-8                    5473            204824 ns/op           4882248 objects/sec    146969 B/op       3065 allocs/op                              
BenchmarkDB_Baseline_Part_Insert-8                                  4282            257142 ns/op           3888905 objects/sec    246580 B/op       4094 allocs/op                              
BenchmarkDB_Baseline_Part_Lookup-8                                 36069             33324 ns/op          30008204 objects/sec         0 B/op          0 allocs/op                              
BenchmarkDB_Baseline_Hashmap_Insert-8                              15986             74292 ns/op          13460333 objects/sec     86561 B/op         64 allocs/op                              
BenchmarkDB_Baseline_Hashmap_Lookup-8                              84038             14228 ns/op          70282379 objects/sec         0 B/op          0 allocs/op                              
BenchmarkDB_DeleteTracker_Baseline-8                                 385           3065639 ns/op            326197 objects/sec   1758394 B/op      23413 allocs/op                              
BenchmarkDB_DeleteTracker-8                                          192           6763615 ns/op            147850 objects/sec   3387030 B/op      40807 allocs/op                              
BenchmarkDB_RandomLookup-8                                          7114            153828 ns/op           6500774 objects/sec       144 B/op          1 allocs/op                              
BenchmarkDB_SequentialLookup-8                                      8168            142537 ns/op           7015728 objects/sec      8000 B/op       1000 allocs/op                              
BenchmarkDB_FullIteration_All-8                                   100930             11497 ns/op          86979024 objects/sec       280 B/op          6 allocs/op                              
BenchmarkDB_FullIteration_Get-8                                    81974             14355 ns/op          69661408 objects/sec       320 B/op          7 allocs/op
BenchmarkDB_PropagationDelay-8                                    243086              4838 ns/op                43.00 50th_µs           50.00 90th_µs          212.0 99th_µs        3083 B/op        50 allocs/op
```

Benchmark after:

```
BenchmarkDB_WriteTxn_1-8                                  380911              3093 ns/op            323322 objects/sec      2440 B/op         38 allocs/op
BenchmarkDB_WriteTxn_10-8                                 901350              1336 ns/op            748741 objects/sec       699 B/op         12 allocs/op
BenchmarkDB_WriteTxn_100-8                               1000000              1075 ns/op            930477 objects/sec       581 B/op          9 allocs/op
BenchmarkDB_WriteTxn_1000-8                              1000000              1262 ns/op            792304 objects/sec       578 B/op          9 allocs/op
BenchmarkDB_WriteTxn_10000-8                             1000000              1262 ns/op            792171 objects/sec       548 B/op          9 allocs/op
BenchmarkDB_WriteTxn_100_SecondaryIndex-8                 705823              1752 ns/op            570791 objects/sec       934 B/op         18 allocs/op
BenchmarkDB_RandomInsert-8                                  1563            743594 ns/op           1344819 objects/sec    481777 B/op       9162 allocs/op
BenchmarkDB_RandomReplace-8                                  530           2266878 ns/op            441135 objects/sec    932101 B/op      19170 allocs/op
BenchmarkDB_SequentialInsert-8                               970           1264837 ns/op            790616 objects/sec    578330 B/op       9168 allocs/op
BenchmarkDB_Baseline_Part_RootOnlyWatch_Insert-8            5121            220673 ns/op           4531599 objects/sec    146970 B/op       3065 allocs/op
BenchmarkDB_Baseline_Part_Insert-8                          4135            281535 ns/op           3551957 objects/sec    246583 B/op       4094 allocs/op
BenchmarkDB_Baseline_Part_Lookup-8                         35305             34346 ns/op          29115518 objects/sec         0 B/op          0 allocs/op
BenchmarkDB_Baseline_Hashmap_Insert-8                      14404             83092 ns/op          12034837 objects/sec     86549 B/op         64 allocs/op
BenchmarkDB_Baseline_Hashmap_Lookup-8                      77744             15222 ns/op          65694567 objects/sec         0 B/op          0 allocs/op
BenchmarkDB_DeleteTracker_Baseline-8                         945           1267367 ns/op            789038 objects/sec    648963 B/op      13062 allocs/op
BenchmarkDB_DeleteTracker-8                                  501           2361545 ns/op            423452 objects/sec   1176999 B/op      20188 allocs/op
BenchmarkDB_RandomLookup-8                                 13999             82787 ns/op          12079196 objects/sec       144 B/op          1 allocs/op
BenchmarkDB_SequentialLookup-8                             14426             77552 ns/op          12894616 objects/sec      8000 B/op       1000 allocs/op
BenchmarkDB_FullIteration_All-8                            86900             12131 ns/op          82435475 objects/sec       264 B/op          6 allocs/op
BenchmarkDB_FullIteration_Get-8                            75100             14046 ns/op          71195485 objects/sec       304 B/op          7 allocs/op
BenchmarkDB_PropagationDelay-8                            398498              3135 ns/op                27.00 50th_µs           32.00 90th_µs          170.0 99th_µs        1479 B/op         28 allocs/op
```

Highlights:
- WriteTxn_1 went from 250k/s to 323k/s
- WriteTxn_100 went from 448k/s to 930k/s
- RandomLookup from 700k/s to 1208k/s
- DeleteTracker up from 150k/s to 423k/s
- Propagation delay down from 43uS to 27uS (50th)
- Reconciler throughput up from ~200k/s to 350k/s, memory usage down 50%

### CODE REVIEW

**Comment by joamaki** [comment]
Benchmark results vs immutable-radix:

```
BenchmarkDB_Baseline_Part_Insert-8                      	   5487	   204858 ns/op	  4881440 objects/sec
BenchmarkDB_Baseline_SingleRadix_Insert-8               	   3091	   375413 ns/op	  2663733 objects/sec
BenchmarkDB_Baseline_SingleRadix_TrackMutate_Insert-8   	   2935	   377532 ns/op	  2648787 objects/sec
BenchmarkDB_Baseline_Part_TrackMutate_Insert-8          	   4450	   255397 ns/op	  3915473 objects/sec
BenchmarkDB_Baseline_SingleRadix_Lookup-8               	  12249	    91621 ns/op	 10914531 objects/sec
BenchmarkDB_Baseline_Part_Lookup-8                      	  12066	    97336 ns/op	 10273722 objects/sec
BenchmarkDB_Baseline_Hashmap_Insert-8                   	  15342	    77111 ns/op	 12968333 objects/sec
BenchmarkDB_Baseline_Hashmap_Lookup-8                   	  83860	    14259 ns/op	 70132148 objects/sec
```

"Part_Insert" is without the watch channel management, and "Part_TrackMutate_Insert" is with. "SingleRadix_TrackMutate" still creates the watch channels but doesn't close them (so it's not apples-to-apples).

**Comment by joamaki** [comment]
The impact of this is really visible in the reconciler benchmark:
```
benchmark % go run . -batchsize 100                                                                                                                                    (pr/joamaki/part)statedb
Inserting batch 1000/1000 ...
Waiting for reconciliation to finish ...

100000 objects reconciled in 0.48 seconds (batch size 100)
Throughput 206555.64 objects per second
Allocated 1423417 objects, 115971kB bytes, 117728kB bytes still in use
benchmark % stg push
Pushing patch "statedb-switch-to-part" ... done (unmodified)
Now at patch "statedb-switch-to-part"
benchmark % go run . -batchsize 100
Inserting batch 1000/1000 ...
Waiting for reconciliation to finish ...

100000 objects reconciled in 0.34 seconds (batch size 100)
Throughput 297510.35 objects per second
Allocated 983554 objects, 57910kB bytes, 60208kB bytes still in use
```

It's 50% faster and we're using almost half the amount of memory for the objects.

File: part/node.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+	"sort"
+	"strings"
+	"unsafe"
+)
+
+type nodeKind uint8
+
+const (
+	nodeKindUnknown = iota
+	nodeKind0
```
Comment (dylandreimerink) [inline part/node.go:18]: What is the point of nodeKind0? It doesn't exists in the paper, and is never set. Any non initialized nodes will be 0 / nodeKindUnknown, so there is also no implicit way for it to exist. Can't we just get rid of it?

File: benchmarks_test.go [hunk 2]
```diff
@@ -197,6 +191,19 @@ func BenchmarkDB_SequentialInsert(b *testing.B) {
...
 
+func BenchmarkDB_Baseline_Part_Insert(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		tree := part.New[uint64](part.RootOnlyWatch)
+		txn := tree.Txn()
+		for j := uint64(0); j < numObjectsToInsert; j++ {
+			txn.Insert(index.Uint64(j), j)
+		}
+		tree = txn.Commit()
+		//require.Equal(b, tree.Len(), numObjectsToInsert)
```
Comment (dylandreimerink) [inline benchmarks_test.go:201]: Do we need to replace with with:
```
if tree.Len() != numObjectsToInsert {
  t.Fatal()
}
```
?

I recall require.Equal being slow in benchmarks, but here its not replaced.

File: benchmarks_test.go [hunk 3]
```diff
@@ -223,6 +230,20 @@ func BenchmarkDB_Baseline_SingleRadix_TrackMutate_Insert(b *testing.B) {
...
+
+func BenchmarkDB_Baseline_Part_TrackMutate_Insert(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		tree := part.New[uint64]()
+		txn := tree.Txn()
+		for j := uint64(0); j < numObjectsToInsert; j++ {
+			txn.Insert(index.Uint64(j), j)
+		}
+		tree = txn.Commit()
+		//require.Equal(b, tree.Len(), numObjectsToInsert)
```
Comment (dylandreimerink) [inline benchmarks_test.go:214]: Ditto

**Review by dylandreimerink — COMMENTED** [review]
Found a few small thing. In general things seem to check out. However, the part/iterator.go and part/txn.go have some very dense code which I find difficult to reason through. That is partially due to also having to wrap my head around the ART implementation which is new to me.

There are a few edge cases in there which are not covered by the current tests, so I would like to schedule a codewalk to go through it together before my final verdict.

File: part/node.go (hunk 1 above)
Comment (joamaki) [reply part/node.go:18]: It's used by the leaf nodes. I'll see about removing it.

File: part/node.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+	watch  chan struct{}
+}
+
+func (n *node[T]) kind() nodeKind {
+	return nodeKind(n.flags >> 12)
+}
+
+func (n *node[T]) setKind(k nodeKind) {
+	n.flags = (n.flags & 0b0000_111_111111111) | (uint16(k) & 0b1111 << 12)
+}
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:43]: Ah yeah, I'll use the same style.

File: benchmarks_test.go (hunk 2 above)
Comment (joamaki) [reply benchmarks_test.go:201]: I'll do a pass to make sure I don't call any `require/assert` in the loops. It's incredibly slow due to it calling `runtime.Callers`. That was also the reason why I called the `callerPackage` thing from `WriteTxn` as it was taking 30% of the time for small WriteTxn's.

**Comment by joamaki** [comment]
> Found a few small thing. In general things seem to check out. However, the part/iterator.go and part/txn.go have some very dense code which I find difficult to reason through. That is partially due to also having to wrap my head around the ART implementation which is new to me.
> 
> There are a few edge cases in there which are not covered by the current tests, so I would like to schedule a codewalk to go through it together before my final verdict.

Thanks for the review! What edge cases are not covered? I think it'd be great to have someone else besides me write a few test cases for this and try to break it...

**Review by bimmlerd — COMMENTED** [review]
`part` looking good so far - need to further review iterator and the tests, and then the integration, but it's hard work :sweat_smile:

File: part/tree.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package part
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/tree.go:4]: Sure! I'll do a pass to add docs to exported types & funcs.

File: part/txn.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+package part
+
+import (
+	"bytes"
+)
+
+type Txn[T any] struct {
+	opts *options
+	root *node[T]
+	size int
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:13]: Number of objects in the tree for implementing `Len()`. I'll document. the fields.

File: part/txn.go [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+	return lowerbound(txn.root, key)
+}
+
+func (txn *Txn[T]) Iterator() *Iterator[T] {
+	clear(txn.mutated)
+	return newIterator[T](txn.root)
+}
+
+func (txn *Txn[T]) Commit() *Tree[T] {
+	txn.mutated = nil
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:90]: Ah interesting. I'll check if `clear` is better from GC perspective!

File: part/part_test.go [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+	_, _, ok := txn.Get(intKey(1))
+	require.False(t, ok)
+}
+
+func Test_delete(t *testing.T) {
+	tree := New[uint64]()
+
+	// Do multiple rounds with the same tree.
+	for round := 0; round < 10; round++ {
+		// Use a random amount of keys in random order to exercise different
```
Comment (bimmlerd) [inline part/part_test.go:83]: can we make this either print the seed or otherwise make it easy to repro the specific random sequence which fails?

File: part/txn.go [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+}
+
+func (txn *Txn[T]) delete(root *node[T], key []byte) (oldValue T, hadOld bool, newRoot *node[T]) {
+	newRoot = root
+	this := root
+
+	// Reuse the same slice in the transaction to hold the parents in order to avoid
+	// allocations.
+	if txn.deleteParents == nil {
+		txn.deleteParents = make([]deleteParent[T], 0, 32)
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:236]: I don't think we need to do anything fancy like that. I just picked a number that's high enough to be most of the time big enough. I can move this into a constant and add some comment about why this number.

File: part/part_test.go [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+
+		// And finally insert the keys back one more time
+		// in random order.
+		rand.Shuffle(len(keys), func(i, j int) {
+			keys[i], keys[j] = keys[j], keys[i]
+		})
+
+		txn = tree.Txn()
+		for _, i := range keys {
+			_, hadOld = txn.Insert(intKey(i), i)
```
Comment (bimmlerd) [inline part/part_test.go:135]: inject another get which confirms they're gone also after committing?

File: part/node.go [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+		n4 := *n.node4()
+		nCopy = (&n4).self()
+	case nodeKind16:
+		n16 := *n.node16()
+		nCopy = (&n16).self()
+	case nodeKind48:
+		n48 := *n.node48()
+		nCopy = (&n48).self()
+	case nodeKind256:
+		nCopy56 := *n.node256()
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:107]: Typo :)

File: part/node.go [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+	if n.leaf != nil {
+		nCopy.leaf = &leaf[T]{
+			key:   n.leaf.key,
+			value: n.leaf.value,
+		}
+	}
+	return nCopy
+}
+
+func (n *node[T]) promote(watch bool) *node[T] {
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:126]: I went for simplicity so I didn't even initially specialize node256 (until I benchmarked and noticed it makes a ton of sense). Since the perf is already 2x to what we had before I would leave it at this for now.

File: part/part_test.go [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+	iter = tree.LowerBound(intKey(3))
+	next(true, 3)
+	next(false, 0)
+
+	iter = tree.LowerBound(intKey(4))
+	next(false, 0)
+}
+
+func Test_iterate(t *testing.T) {
+	sizes := []int{0, 1, 10, 100, 1000, rand.Intn(1000)}
```
Comment (bimmlerd) [inline part/part_test.go:387]: any specific behaviour we want for calling Next() on iterators which already finished?

File: part/node.go [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+func (n *node[T]) children() []*node[T] {
+	switch n.kind() {
+	case nodeKind4:
+		return n.node4().children[0:n.size():4]
+	case nodeKind16:
+		return n.node16().children[0:n.size():16]
+	case nodeKind48:
+		return n.node48().children[0:n.size():48]
+	case nodeKind256:
+		return n.node256().children[:]
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:205]: We need `[:]` to convert it to a slice. `[:256]` would be the same as `[:]`.

File: part/iterator.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+		} else if len(edges) > 1 {
+			// More edges remain to be explored, add them back.
+			it.next = append(it.next, edges[1:])
+		}
+
+		// Follow the smallest edge and add its children to the queue.
+		node := edges[0]
+
+		if node.size() > 0 {
+			if node.kind() == nodeKind256 {
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/iterator.go:164]: Yeah we don't. Leftover from when I was doing the conversion to the optimized `Node256` and I made `children()` panic for Node256 so I would catch everything. Fixing.

File: part/iterator.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+	return &Iterator[T]{[][]*node[T]{{start}}}
+}
+
+func lowerbound[T any](start *node[T], key []byte) *Iterator[T] {
+	// The starting edges to explore. This contains all larger nodes encountered
+	// on the path to the node larger or equal to the key.
+	edges := [][]*node[T]{}
+
+	var traverseToMin func(n *node[T])
+	traverseToMin = func(n *node[T]) {
```
Comment (bimmlerd) [inline part/iterator.go:53]: nit: definition and declaration separate, could be joined?

File: part/node.go [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	nodeKind48
+	nodeKind256
+)
+
+type leaf[T any] struct {
+	key   []byte
+	value T
+}
+
+type node[T any] struct {
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:29]: If it uses less memory then we should go for it. Reordered.

File: part/txn.go [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+			// Update target to point to the cloned node
+			children[target.index] = target.node
+		}
+
+		if parent.node.size() > 0 {
+			// Check if the node should be demoted.
+			// TODO: Can we avoid the initial clone of parent.node?
+			var newNode *node[T]
+			switch {
+			case parent.node.kind() == nodeKind256 && parent.node.size() <= 48:
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:322]: If we don't demote at the border we would be wasting space and then we might as well make the node a different size?

How would the underlying array be retained (it's embedded in the node) and wouldn't it be bad to retain it since we would then use more memory?

**Review by bimmlerd — APPROVED** [review]
LGTM pending a few nits and questions, but I think this is good enough as is already.

File: part/txn.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:90]:
```
goos: linux
goarch: amd64
pkg: github.com/cilium/statedb/part
cpu: Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
Benchmark_txn_1-8        1608686               775.2 ns/op         1290049 objects/sec
PASS
ok      github.com/cilium/statedb/part  2.275s
part % go test . -bench txn_1\$                                                     (pr/joamaki/part)statedb
goos: linux
goarch: amd64
pkg: github.com/cilium/statedb/part
cpu: Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
Benchmark_txn_1-8        1767748               643.4 ns/op         1554270 objects/sec
PASS
```

First one is with `clear`, second without. So clearly worse.

File: part/part_test.go (hunk 13 above)
Comment (joamaki) [reply part/part_test.go:483]: Added assert.

File: part/iterator.go (hunk 16 above)
Comment (joamaki) [reply part/iterator.go]: No it can't be since `traverseToMin` calls `traverseToMin`

Comment (joamaki) [reply part/iterator.go]: I lifted `traverseToMin` into a top-level function.

File: part/part_test.go (hunk 10 above)
Comment (joamaki) [reply part/part_test.go:141]: Done.

File: part/txn.go [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	for i := len(parents) - 1; i > 0; i-- {
+		parent := &parents[i-1]
+		target := &parents[i]
+
+		// Clone the parent to mutate it.
+		parent.node = txn.cloneNode(parent.node)
+		var children []*node[T]
+		if parent.node.kind() == nodeKind256 {
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go:321]: Removed. Difference is that `.children[:]` is a slice type and `.children` is an array type.

File: part/txn.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (joamaki) [reply part/txn.go]: I set the demoting to be at 37, 12 and 3 to avoid the thrashing.

File: part/node.go [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+		copy(node48.children[:], node16.children[:node16.size()])
+		if watch {
+			node48.watch = make(chan struct{})
+		}
+		return node48.self()
+	case nodeKind48:
+		node48 := n.node48()
+		node256 := &node256[T]{node: *n}
+		node256.setKind(nodeKind256)
+		for _, child := range node48.children[:node48.size()] {
```
[the remark this answers is not in the document]
Comment (joamaki) [reply part/node.go:195]: Done.

File: part/txn.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bimmlerd) [reply part/txn.go:116]: `clear`ly worse :grin: anyway - good to know!

**Review by pippolo84 — APPROVED** [review]
Focused my review on part and looks good to me. :rocket: 
I'll take another look at the benchmarks and tests later, but don't need to block this further.
